src/JSON.jl

```diff
@@ -68,116 +131,88 @@ function end_object(io::IO, state::State{NOINDENT}, is_dict::Bool)
-    function _print(io::IO, state::State, s::Base.Dates.TimeType)
-        _print(io, state, string(s))
+function print_escaped(io::IO, s::Compat.UTF8String)
+    @inbounds for c in s.data
+        Base.write(io, escaped[c + 0x01])
     end
 end
 
-function _print(io::IO, state::State, s::Symbol)
-    _print(io, state, string(s))
+function _writejson(io::IO, state::State, s::AbstractString)
+    Base.print(io, '"')
+    JSON.print_escaped(io, s)
+    Base.print(io, '"')
 end
 
-@compat function _print(io::IO, state::State, s::Union{Integer, AbstractFloat})
+@compat function _writejson(io::IO, state::State, s::Union{Integer, AbstractFloat})
     if isnan(s) || isinf(s)
         Base.print(io, "null")
     else
         Base.print(io, s)
     end
 end
 
-@compat function _print(io::IO, state::State, n::Void)
+@compat function _writejson(io::IO, state::State, n::Void)
     Base.print(io, "null")
 end
 
-function _print(io::IO, state::State, a::Associative)
+function _writejson(io::IO, state::State, a::Associative)
     if length(a) == 0
         Base.print(io, "{}")
         return
     end
     start_object(io, state, true)
     first = true
-    for (key, value) in a
+    for key in keys(a)
         first ? (first = false) : Base.print(io, ",", suffix(state))
         Base.print(io, prefix(state))
-        JSON._print(io, state, string(key))
+        _writejson(io, state, string(key))
         Base.print(io, separator(state))
-        JSON._print(io, state, value)
+        _writejson(io, state, a[key])
     end
     end_object(io, state, true)
 end
 
-@compat function _print(io::IO, state::State, a::Union{AbstractVector,Tuple})
+@compat function _writejson(io::IO, state::State, a::Union{AbstractVector,Tuple})
     if length(a) == 0
         Base.print(io, "[]")
         return
     end
     start_object(io, state, false)
     Base.print(io, prefix(state))
     i = start(a)
-    !done(a,i) && ((x, i) = next(a, i); JSON._print(io, state, x); )
+    !done(a,i) && ((x, i) = next(a, i); _writejson(io, state, x); )
 
     while !done(a,i)
         (x, i) = next(a, i)
         Base.print(io, ",")
         printsp(io, state)
-        JSON._print(io, state, x)
+        _writejson(io, state, x)
     end
     end_object(io, state, false)
 end
 
-function _print(io::IO, state::State, a)
-    start_object(io, state, true)
-    range = @compat fieldnames(a)
-    if length(range) > 0
-        Base.print(io, prefix(state), "\"", range[1], "\"", separator(state))
-        JSON._print(io, state, getfield(a, range[1]))
-
-        for name in range[2:end]
-            Base.print(io, ",")
-            printsp(io, state)
-            Base.print(io, "\"", name, "\"", separator(state))
-            JSON._print(io, state, getfield(a, name))
-        end
+function _writejson(io::IO, state::State, a)
+    # FIXME: This fallback is harming performance substantially.
+    # Remove this fallback when _print removed.
+    if applicable(_print, io, state, a)
```

Comment:
Caching the results in a `Dict` only saves a tiny portion of the runtime cost, while using more memory, adding compile-time overhead, and also making the entire thing not thread safe. I think we might have to live with the `applicable` check for now.